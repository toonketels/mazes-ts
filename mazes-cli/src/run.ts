import {createStringPainter, paintMazeCreation, wallPainter} from "mazes-lib";
import {clearScreen} from "ansi-escapes";
import {Dimensions} from "mazes-lib/dist/lib/grid";

export interface RunOptions {
    dimensions: Dimensions,
    delayMs?: number,
    seed?: string
}

export async function run({dimensions, delayMs = 250, seed}: RunOptions): Promise<void> {

    return new Promise(resolve => {

        const frames = paintMazeCreation(withOptions(dimensions, seed));

        const interval = setInterval(() => {

            let {done, value: frame} = frames.next()

            if (done) {
                clearInterval(interval)
                resolve()
            }

            process.stdout.write(`${clearScreen}${frame}`)

        }, delayMs);
    });
}

function withOptions(dimensions: Dimensions, seed: string | undefined) {
    let creationOptions = {
        dimensions,
        painter: createStringPainter({cellPainter: wallPainter})
    };
    return {...creationOptions, ...(seed && {seed})}
}
