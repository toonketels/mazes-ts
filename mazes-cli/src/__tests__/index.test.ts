import {run} from "../run";

describe("cli", () => {

    let spy: jest.SpyInstance

    beforeAll(() => {
        spy = jest.spyOn(process.stdout, 'write').mockImplementation(function () { return true; });
    });

    afterAll(() => {
        jest.restoreAllMocks()
    })

    test("animate the maze building to std out", async () => {

        await run({dimensions: {width: 3, height: 3}, seed: "stabilize the input", delayMs: 0})

        expect(process.stdout.write).toHaveBeenCalledTimes(11)
        for (let [arg] of spy.mock.calls) expect(arg).toMatchSnapshot()
    })

})