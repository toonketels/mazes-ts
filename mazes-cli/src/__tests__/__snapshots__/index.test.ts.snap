// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli animate the maze building to std out 1`] = `
"c+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 2`] = `
"c+---+---+---+
|       |   |
+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 3`] = `
"c+---+---+---+
|       |   |
+---+   +---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 4`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 5`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|       |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 6`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|           |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 7`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|           |
+---+---+   +
|   |   |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 8`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|           |
+---+---+   +
|       |   |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 9`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|           |
+---+---+   +
|           |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 10`] = `
"c+---+---+---+
|       |   |
+---+   +   +
|           |
+---+---+   +
|           |
+---+---+---+
"
`;

exports[`cli animate the maze building to std out 11`] = `"cundefined"`;
